dao/src/main/scala/za/co/absa/enceladus/dao/rest/RestClient.scala

```diff
@@ -76,23 +83,37 @@ protected class RestClient(authClient: AuthClient,
 
     val statusCode = response.getStatusCode
 
+    /**
+     * This function handles unauthorized response by trying to authenticate
+     * (if there are still some retries attempt left) - this might be due to an expired session.
+     *
+     * @param exceptionToThrow Instance of an exception that should be thrown if there are no retries left.
+     */
+    def handleUnauthorizedResponse(exceptionToThrow: OptionallyRetryableException): Unit = {
+      log.warn(s"Response - ${statusCode} : ${Option(response.getBody).getOrElse("<empty body>")}")
+      log.warn(s"Unauthorized $method request for Menas URL: $url")
+      if (retriesLeft <= 0) {
+        throw exceptionToThrow
+      }
+
+      log.warn(s"Expired session, reauthenticating")
+      authenticate()
+
+      log.info(s"Retrying $method request for Menas URL: $url")
+      log.info(s"Retries left: $retriesLeft")
+    }
+
     statusCode match {
       case HttpStatus.OK | HttpStatus.CREATED             =>
         log.info(s"Response (${response.getStatusCode}): ${response.getBody}")
         JsonSerializer.fromJson[T](response.getBody)
 
-      case HttpStatus.UNAUTHORIZED | HttpStatus.FORBIDDEN =>
-        log.warn(s"Response - $statusCode : ${Option(response.getBody).getOrElse("None")}")
-        log.warn(s"Unauthorized $method request for Menas URL: $url")
-        if (retriesLeft <= 0) {
-          throw UnauthorizedException("Unable to reauthenticate, no retries left")
-        }
-
-        log.warn(s"Expired session, reauthenticating")
-        authenticate()
+      case HttpStatus.UNAUTHORIZED =>
+        handleUnauthorizedResponse(UnauthorizedException("Unauthorized, unable to reauthenticate, no retries left"))
+        send[B, T](method, url, headers, bodyOpt, retriesLeft - 1)
```

Comment:
Why is the `send` not part of the `handleUnauthorizedResponse`?

Reply:
I tried but it wouldn't be a tail recursive function - at least I couldn't make it work

Reply:
Indeed, I missed the fact that the function is recursive. 
It's good to comment such things - why some **obvious* thing was **not done*, similarly why some **not obvious** things **were done** 😉